src/model.ts

```diff
@@ -1,3 +1,44 @@
-export interface User {
-  name: string;
+export interface ValidationResult {
+  key?: string;
+  type: string;
+  succeeded: boolean;
+  errorMessage: string;
+}
+
+export const createDefaultValidationResult = (): ValidationResult => ({
+  key: '',
+  type: '',
+  succeeded: true,
+  errorMessage: '',
+});
+
+// Pending on ValidationEngine
+type ValidationResultSyncAsync = ValidationResult | Promise<ValidationResult>;
+
+export type FieldValidationFunction = (
+  value: any,
+  values: any,
+  customParams: any
+) => Promise<ValidationResult>;
+
+// renamoe to FullFieldValidation
+export interface FullFieldValidation {
+  validator: FieldValidationFunction;
+  customArgs?: object;
+  errorMessage: string | string[];
+}
+
+export type FieldValidation = FieldValidationFunction | FullFieldValidation;
+
+export interface FormValidation {
+  succeeded: boolean;
+  fieldErrors: { [key: string]: ValidationResult };
+  formGlobalErrors: Array<ValidationResult>;
+}
+
+export type RecordValidationFunction = (vm: any) => ValidationResult;
```

Comment:
The function param is `values` ?

Reply:
It should return a Promise<ValidationResult>?

Reply:
You are right, still work to do on this (once I hop into the Record Validation implementation), I think the signature for this function should be promise or just the validation result (we accept both flavours).

Just refactoring to your proposal and later on will do some tweaking once jumping into that case.